client/rpc/client.go

```diff
@@ -0,0 +1,164 @@
+package rpc
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"math/rand"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/nats-io/nats.go"
+	"github.com/rs/zerolog"
+	"github.com/statechannels/go-nitro/client/engine/store/safesync"
+	"github.com/statechannels/go-nitro/network"
+	"github.com/statechannels/go-nitro/network/serde"
+	natstrans "github.com/statechannels/go-nitro/network/transport/nats"
+	"github.com/statechannels/go-nitro/protocols"
+	"github.com/statechannels/go-nitro/protocols/directdefund"
+	"github.com/statechannels/go-nitro/protocols/directfund"
+	"github.com/statechannels/go-nitro/protocols/virtualfund"
+	"github.com/statechannels/go-nitro/types"
+
+	"github.com/statechannels/go-nitro/channel/state/outcome"
+)
+
+// RpcClient is a client for making nitro rpc calls
+type RpcClient struct {
+	nts       *network.NetworkService
+	myAddress types.Address
+	chainId   *big.Int
+
+	// responses is a collection of channels that are used to wait until a response is received from the RPC server
+	responses safesync.Map[chan interface{}]
+
+	idsToMethods safesync.Map[serde.RequestMethod]
+}
+
+// NewRpcClient creates a new RpcClient
+func NewRpcClient(rpcServerUrl string, myAddress types.Address, chainId *big.Int, logger zerolog.Logger) *RpcClient {
+
+	nc, err := nats.Connect(rpcServerUrl)
+	handleError(err)
+	trp := natstrans.NewNatsTransport(nc, getTopics())
+
+	con, err := trp.PollConnection()
+	handleError(err)
+	nts := network.NewNetworkService(con)
+	nts.Logger = logger
+	c := &RpcClient{nts, myAddress, chainId, safesync.Map[chan interface{}]{}, safesync.Map[serde.RequestMethod]{}}
+	c.registerHandlers()
+	return c
+}
+
+// CreateLedger creates a new ledger channel
+func (rc *RpcClient) CreateLedger(counterparty types.Address, ChallengeDuration uint32, outcome outcome.Exit) directfund.ObjectiveResponse {
+
+	objReq := directfund.NewObjectiveRequest(
+		counterparty,
+		100,
+		outcome,
+		uint64(rand.Float64()), // TODO: Since numeric fields get converted to a float64 in transit we need to prevent overflow
+		common.Address{})
+
+	// Create a channel and store it in the responses map
+	// We will use this channel to wait for the response
+	resRec := make(chan interface{})
```

Comment:
Is there scope for DRYing out some of the code common to the various methods?

Reply:
This was pointed out in a comment in a currently open followup PR https://github.com/statechannels/go-nitro/pull/1084/files#r1094726558. Probably best to tackle DRYing out refactor in [that PR](https://github.com/statechannels/go-nitro/pull/1084) as that PR builds off this code.